compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -898,34 +1245,53 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
             }
         }
 
-        ty::FnDef(def_id, args) => {
-            ty = Ty::new_fn_def(tcx, *def_id, transform_args(tcx, args, options));
+        ty::FnDef(def_id, ..) => {
+            // Transform function items into function pointers
+            let fn_sig = tcx.fn_sig(def_id).skip_binder();
+            let fn_ptr = Ty::new_fn_ptr(tcx, fn_sig);
+            // Transform fn_sig inputs and output
+            ty = transform_ty(tcx, fn_ptr, options);
         }
 
-        ty::Closure(def_id, args) => {
-            ty = Ty::new_closure(tcx, *def_id, transform_args(tcx, args, options));
+        ty::Closure(_, args) => {
+            // Transform closures into function pointers
+            let fn_sig = tcx.signature_unclosure(args.as_closure().sig(), hir::Unsafety::Normal);
+            let fn_ptr = Ty::new_fn_ptr(tcx, fn_sig);
+            // Transform fn_sig inputs and output
+            ty = transform_ty(tcx, fn_ptr, options);
         }
 
-        ty::CoroutineClosure(def_id, args) => {
-            ty = Ty::new_coroutine_closure(tcx, *def_id, transform_args(tcx, args, options));
-        }
-
-        ty::Coroutine(def_id, args) => {
-            ty = Ty::new_coroutine(tcx, *def_id, transform_args(tcx, args, options));
+        ty::Coroutine(_, args) => {
+            let def_id = tcx.require_lang_item(hir::LangItem::Coroutine, None);
+            let gen_sig = args.as_coroutine().sig();
+            ty = new_dynamic_trait(
+                tcx,
+                def_id,
+                tcx.mk_args(&[
+                    gen_sig.resume_ty.into(),
+                    gen_sig.yield_ty.into(),
+                    gen_sig.return_ty.into(),
+                ]),
+            );
         }
 
         ty::Ref(region, ty0, ..) => {
-            if options.contains(TransformTyOptions::GENERALIZE_POINTERS) {
-                if ty.is_mutable_ptr() {
-                    ty = Ty::new_mut_ref(tcx, tcx.lifetimes.re_static, Ty::new_unit(tcx));
-                } else {
-                    ty = Ty::new_imm_ref(tcx, tcx.lifetimes.re_static, Ty::new_unit(tcx));
-                }
+            // Remove references from function items, closures, and Fn trait objects
+            if ty0.is_fn_def() || ty0.is_closure() || is_dynamic_fn_trait(tcx, *ty0) {
+                ty = transform_ty(tcx, *ty0, options);
             } else {
-                if ty.is_mutable_ptr() {
-                    ty = Ty::new_mut_ref(tcx, *region, transform_ty(tcx, *ty0, options));
+                if options.contains(TransformTyOptions::GENERALIZE_POINTERS) {
+                    if ty.is_mutable_ptr() {
```

Comment:
I believe it makes more sense to be turning `&mut T` into `*mut T` and not the other way around.

Reply:
In this case, we're testing the mutability of a ty::Ref (not ty:RawPtr). It just so happens that the name of the helper function from Sty is called `is_mutable_ptr` (but it tests mutability for both ty::Ref and ty::RawPtr).